### ISSUE TITLE
Add TPU/multi-GPU support for ensembles

### ISSUE THREAD

**Issue by dustinvtran** [google/uncertainty-baselines#63]
> @dustinvtran 
> @GhassenJ @znado @JasperSnoek
> 
> To do this, we need to replace storing logits/labels in memory with an eval loop.
> 
> The simplest method is that at each iteration, each worker loops over checkpoints to grab logit predictions, then aggregates to make the ensemble prediction, and then runs eval against the current batch's labels. However, loading checkpoints at each iteration may be prohibitively expensive.
> 
> An alternative method is to 1. write logits in parallel to a file using the workers; 2. in the coordinator, aggregate the logits and eval against the dataset's labels which is batched once more. This second step should be inexpensive compared to the first step; the second step does not run the model and does not require image data in memory.
> 
> Here's pseudocode.
> 
> ```python
> for checkpoint, logits_file in zip(checkpoints, logits_files):
>   model.load(checkpoint)
>   for batch in dataset:  # the dataset is split in parallel across workers
>     predict_fn(model, batch, logits_file)  # predict_fn calls model and writes logits to logits_file
> 
> # Perform below in coordinator (CPU).
> for batch in dataset:
>   logits = [tf.nn.softmax(subset(logits_file, batch)) for logits_file in logits_files]
>   probs = tf.reduce_mean(logits, axis=-1)
>   eval(batch[labels], probs)
> ```
> 
> A final method involves having each worker load different checkpoints. However, this requires model parallelism in addition to data parallelism, which is not supported well.
>
> Partially resolved in https://github.com/google/edward2/pull/258. The next step is to figure out exactly how to parallelize writing logits to files.
> 
> In general, there are M models, N datasets, and M*N files of logits.
> 
> * It's easy to parallelize writing logits files across models or datasets, but only via separate jobs rather than a single job. Separate jobs can be fine, but this makes the pipeline more complex as you need to manually run the second step of eval that reads from all these files.
> * If we restrict to a single job, we must resort to data parallelism, so somehow asynchronously writing logits to files while still preserving the dataset ordering. (This will also not be storeable as a NumPy array unless the workers all communicate their predictions to the coordinator, which is very expensive. So it may have to be something like plaintext?)

https://github.com/google/edward2/issues/249

**Comment by dustinvtran** [comment]
This is resolved with Robustness Metrics' pipeline (to be open-sourced). I have yet to document the precise steps.
